include/h2/tensor/hydrogen_interop/dist_tensor_interop.hpp

```diff
@@ -0,0 +1,492 @@
+{
+    switch (d)
+    {
+    case El::MC:
+    case El::MR:
+    case El::VC:
+    case El::VR: return Distribution::Block;
+    case El::STAR: return Distribution::Replicated;
+    case El::CIRC: return Distribution::Single;
+    default: return Distribution::Undefined;
+    }
+}
+
+inline El::mpi::Comm const&
+logical_1d_comm(El::Grid const& g, El::Dist coldist, El::Dist rowdist) noexcept
+{
+    if ((coldist == El::MR && rowdist == El::MC)
+        || (coldist == El::MR && rowdist == El::STAR)
+        || (coldist == El::STAR && rowdist == El::MC)
+        || (coldist == El::STAR && rowdist == El::VR)
+        || (coldist == El::VR && rowdist == El::STAR))
+    {
+        return g.VRComm();
+    }
+    return g.VCComm();
+}
+
+inline ShapeTuple
+make_canonical_grid_shape(El::Grid const& g, El::Dist coldist, El::Dist rowdist)
+{
+#define MATCHDIST_RETURN_VAL(CDIST, RDIST, ...)                                \
+    do                                                                         \
+    {                                                                          \
+        if (coldist == CDIST && rowdist == RDIST)                              \
+        {                                                                      \
+            return __VA_ARGS__;                                                \
+        }                                                                      \
+    } while (0)
+
+    using namespace El::DistNS;
+    MATCHDIST_RETURN_VAL(MC, MR, {g.MCSize(), g.MRSize()});
+    MATCHDIST_RETURN_VAL(MC, STAR, {g.MCSize(), g.MRSize()});
+    MATCHDIST_RETURN_VAL(MR, MC, {g.MRSize(), g.MCSize()});
+    MATCHDIST_RETURN_VAL(MR, STAR, {g.MRSize(), g.MCSize()});
+    MATCHDIST_RETURN_VAL(STAR, MC, {g.MRSize(), g.MCSize()});
+    MATCHDIST_RETURN_VAL(STAR, MR, {g.MCSize(), g.MRSize()});
+    MATCHDIST_RETURN_VAL(STAR, VC, {1, g.VCSize()});
+    MATCHDIST_RETURN_VAL(STAR, VR, {1, g.VRSize()});
+    MATCHDIST_RETURN_VAL(VC, STAR, {g.VCSize(), 1});
+    MATCHDIST_RETURN_VAL(VR, STAR, {g.VRSize(), 1});
+    MATCHDIST_RETURN_VAL(STAR, STAR, {g.VCSize(), 1});
+    MATCHDIST_RETURN_VAL(CIRC, CIRC, {g.VCSize(), 1});
+#undef MATCHDIST_RETURN_VAL
+    throw std::logic_error("Unknown Hydrogen distribution.");
+}
+
+inline ProcessorGrid
+make_default_grid(El::Grid const& g, El::Dist coldist, El::Dist rowdist)
+{
+    auto const& comm = logical_1d_comm(g, coldist, rowdist);
+    return ProcessorGrid{comm, make_canonical_grid_shape(g, coldist, rowdist)};
+}
+
+template <El::Device D, typename T, typename U>
+auto make_dist_tensor_impl(T* const buffer,
+                           El::AbstractDistMatrix<U> const& mat)
+{
+    static_assert(meta::Eq<std::decay_t<T>, U>);
+    auto const coldist = mat.ColDist();
+    auto const rowdist = mat.RowDist();
+    return DistTensor<U>{
+        H2Device<D>,
+        buffer,
+        ShapeTuple{safe_as<int>(mat.Height()), safe_as<int>(mat.Width())},
```

Comment:
The underlying datatype for `ShapeTuple` is not necessarily an `int`, maybe this should be `safe_as<ShapeTuple::type>`?